ext/js/language/language-transformer.js

```diff
@@ -120,18 +121,25 @@ export class LanguageTransformer {
         for (let i = 0; i < results.length; ++i) {
             const {text, conditions, trace} = results[i];
             for (const transform of this._transforms) {
-                if (!transform.suffixHeuristic.test(text)) { continue; }
+                if (!transform.heuristic.test(text)) { continue; }
 
                 const {name, rules} = transform;
                 for (let j = 0, jj = rules.length; j < jj; ++j) {
                     const rule = rules[j];
                     if (!LanguageTransformer.conditionsMatch(conditions, rule.conditionsIn)) { continue; }
-                    const {suffixIn, suffixOut} = rule;
-                    if (!text.endsWith(suffixIn) || (text.length - suffixIn.length + suffixOut.length) <= 0) { continue; }
+                    const {isInflected, uninflect} = rule;
+                    if (!isInflected.test(text)) { continue; }
+
+                    const isCycle = trace.some((frame) => frame.transform === name && frame.ruleIndex === j && frame.text === text);
+                    if (isCycle) {
+                        log.warn(new Error(`Cycle detected in transform[${name}] rule[${j}] for text: ${text}`));
+                        continue;
+                    }
+
                     results.push(LanguageTransformer.createTransformedText(
-                        text.substring(0, text.length - suffixIn.length) + suffixOut,
+                        uninflect(text),
```

Comment:
Shouldn't this be `deinflect` as opposed to `uninflect`, to fit the other parts of the code? Unless there is some quirky difference in meaning between `uninflect` and `deinflect` that I am not getting. <!-- :10catwhaa: -->

Reply:
Back in my day, deinflect was a function in the Deinflector... yeah I should change that now